### PR TITLE
SBERDOMA-378 First iteration of tables DRY

### PR DESCRIPTION
**Pull request by SavelevMatthew**
For more details refer to this article in confluence: https://doma.atlassian.net/wiki/spaces/SBER/pages/787808257/1

### CODE REVIEW

**Review by Dimitreee — APPROVED** [review]

**Comment by SavelevMatthew** [comment]
just rebasing, nothing changed since last review

File: apps/condo/domains/billing/components/BillingPageContent/ReceiptsTable.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+} from '@condo/domains/common/utils/tables.utils'
+import { useQueryMappers } from '@condo/domains/common/hooks/useQueryMappers'
+import { useRouter } from 'next/router'
+import { useIntl } from '@core/next/intl'
+import { Table, DEFAULT_PAGE_SIZE } from '@condo/domains/common/components/Table/Index'
+import { BillingReceipt } from '@condo/domains/billing/utils/clientSchema'
+import { SortBillingReceiptsBy } from '../../../../schema'
+import get from 'lodash/get'
+import { useSearch } from '@condo/domains/common/hooks/useSearch'
+import { usePeriodSelector } from '../../hooks/usePeriodSelector'
```
Comment (Dimitreee) [inline apps/condo/domains/billing/components/BillingPageContent/ReceiptsTable.tsx:21]: condo

File: apps/condo/domains/billing/hooks/usePeriodSelector.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import { useRouter } from 'next/router'
+import { parseQuery } from '../../common/utils/tables.utils'
```
Comment (Dimitreee) [inline apps/condo/domains/billing/hooks/usePeriodSelector.ts:2]: @condo

File: apps/condo/domains/common/components/Table/Index.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    const handleChange = debounce((...tableChangeArguments) => {
+        const [
+            nextPagination,
+            nextFilters,
+            nextSorters,
+        ] = tableChangeArguments
+        const { current } = nextPagination
+        let shouldResetOffset = false
+
+        const newFilters: { [x: string]: QueryArgType } = { ...filters }
```
Comment (Dimitreee) [inline apps/condo/domains/common/components/Table/Index.tsx:53]: move to typedefs

Comment (Dimitreee) [reply apps/condo/domains/common/components/Table/Index.tsx:53]: add readable prop name

File: apps/condo/domains/common/components/Table/Index.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        for (const [key, value] of Object.entries(nextFilters)) {
```
Comment (Dimitreee) [inline apps/condo/domains/common/components/Table/Index.tsx:54]: add readable key/value names

File: apps/condo/domains/common/hooks/useQueryMappers.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        }, [])
+        const validMetas = queryMetas
+            .filter((meta) => meta && meta.keyword && meta.filters && meta.filters.length > 0)
+            .map((meta) => ({ [meta.keyword]: meta }))
+
+        const filtersMap = Object.assign({}, ...validMetas)
+
+        const filtersToWhere = (queryFilters) => {
+            const whereQueries = []
+            for (const [key, value] of Object.entries(queryFilters)) {
```
Comment (Dimitreee) [inline apps/condo/domains/common/hooks/useQueryMappers.ts:18]: add readable names

File: apps/condo/domains/common/hooks/useQueryMappers.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+            return {
+                AND: whereQueries,
+            }
+        }
+
+        const sortersToSortBy = (querySorts: SorterColumn | Array<SorterColumn>) => {
+            const sorters = convertSortersToSortBy(querySorts)
+                .filter((sortLine) => validSorts.includes(sortLine))
+            return sorters.length ? sorters : ['createdAt_DESC']
```
Comment (Dimitreee) [inline apps/condo/domains/common/hooks/useQueryMappers.ts:37]: const DEFAULT_SORT = 'createdAt_DESC'

File: apps/condo/domains/common/utils/tables.utils.ts [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+) => FilterType = (
+    dataIndex, argType, argData, suffix
+) => {
+    let wrappedDataIndex = []
+    if (Array.isArray(dataIndex) && dataIndex.every(Boolean)) {
+        wrappedDataIndex = dataIndex
+    } else if (typeof dataIndex === 'string' && dataIndex) {
+        wrappedDataIndex = [dataIndex]
+    }
+    return function (search) {
```
Comment (Dimitreee) [inline apps/condo/domains/common/utils/tables.utils.ts:109]: why anonymus function?

File: apps/condo/domains/common/utils/tables.utils.ts [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    return getFilter(dataIndex, 'single', 'number', null)
+}
+
+export const getStringOptionFilter: (dataIndex: DataIndexType) => FilterType = (dataIndex) => {
+    return getFilter(dataIndex, 'array', 'string', 'in')
+}
+
+export const getDayGteFilter: (dataIndex: DataIndexType) => FilterType = (dataIndex) => {
+    const filter = getFilter(dataIndex, 'single', 'dateTime', 'gte')
+    return function (search) {
```
Comment (Dimitreee) [inline apps/condo/domains/common/utils/tables.utils.ts:165]: why anonymus function?

File: apps/condo/domains/common/utils/tables.utils.ts [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    if (!filters || typeof filters !== 'string') {
+        return {}
+    }
+    try {
+        const json = JSON.parse(filters)
+        const result = {}
+        Object.keys(json).forEach((key) => {
+            const value = json[key]
+            if (Array.isArray(value)) {
+                result[key] = value.filter((v) => typeof v === 'string' && !!v)
```
Comment (Dimitreee) [inline apps/condo/domains/common/utils/tables.utils.ts:228]:
```suggestion
                result[key] = value.filter((v) => typeof v === 'string' && Boolean(v))
```

File: apps/condo/domains/common/utils/tables.utils.ts [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+    const filters = getFiltersFromQuery(query)
+    const sorters = getSortersFromQuery(query)
+    const queryOffset = get(query, 'offset', '0')
+    const offset = Number(queryOffset) ? Number(queryOffset) : 0
+    return { filters, sorters, offset }
+}
+
+export const convertColumns = (
+    columns: Array<ColumnInfo>,
+    filters: { [x: string]: QueryArgType },
```
Comment (Dimitreee) [inline apps/condo/domains/common/utils/tables.utils.ts:253]: move to types, add readable variabale names

File: apps/condo/domains/billing/components/BillingPageContent/ReceiptsTable.tsx (hunk 1 above)
Comment (SavelevMatthew) [reply apps/condo/domains/billing/components/BillingPageContent/ReceiptsTable.tsx]: fixed 👍

File: apps/condo/domains/billing/hooks/usePeriodSelector.ts (hunk 2 above)
Comment (SavelevMatthew) [reply apps/condo/domains/billing/hooks/usePeriodSelector.ts]: fixed 👍

File: apps/condo/domains/common/hooks/useQueryMappers.ts (hunk 5 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/hooks/useQueryMappers.ts]: added and styled this function to look more linear

File: apps/condo/domains/common/components/Table/Index.tsx (hunk 3 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/components/Table/Index.tsx:53]: done 👍

File: apps/condo/domains/common/components/Table/Index.tsx (hunk 4 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/components/Table/Index.tsx:54]: done 👍

File: apps/condo/domains/common/hooks/useQueryMappers.ts (hunk 6 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/hooks/useQueryMappers.ts]: done 👍

File: apps/condo/domains/common/utils/tables.utils.ts (hunk 7 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/utils/tables.utils.ts:109]: refactored 👍

File: apps/condo/domains/common/utils/tables.utils.ts (hunk 8 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/utils/tables.utils.ts:165]: refactored 👍

File: apps/condo/domains/common/utils/tables.utils.ts (hunk 9 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/utils/tables.utils.ts:228]: refactored 👍

File: apps/condo/domains/common/utils/tables.utils.ts (hunk 10 above)
Comment (SavelevMatthew) [reply apps/condo/domains/common/utils/tables.utils.ts:253]: done 👍

**Review by sitozzz — APPROVED** [review]
LGTM

**Pull request merged**